interface/modules/custom_modules/oe-module-faxsms/src/RCVoice/VoiceFunctionsTrait.php

```diff
@@ -0,0 +1,527 @@
+            http_response_code(500);
+            header('Content-Type: application/json');
+            return json_encode(['error' => $e->getMessage()]);
+        }
+    }
+
+
+    /**
+     * Initialize the Platform instance (call from your constructor).
+     */
+    protected function initVoice($platform): void
+    {
+        $this->platform = $platform;
+        $this->install();
+    }
+
+    /**
+     * Make an outbound call (RingOut).
+     */
+    public function makeRingOutCall($toNumber = null, $fromNumber = null, $callerId = null)
+    {
+        $authErrorMsg = $this->authenticate();
+        if ($authErrorMsg !== 1) {
+            return text(js_escape($authErrorMsg));
+        }
+
+        $toNumber = $toNumber ?: $this->getRequest('toPhone');
+        $fromNumber = $fromNumber ?: $this->getRequest('fromPhone');
+        $userNumber = $this->credentials['smsNumber'] ?? $this->credentials['phone'] ?? '';
+        $defaultNumber = $this->formatPhone($userNumber);
+        $fromNumber = $this->formatPhone($fromNumber ?: $defaultNumber);
+        $toNumber = $this->formatPhone($toNumber);
+
+        try {
+            $body = [
+                'from' => ['phoneNumber' => $fromNumber],
+                'to' => ['phoneNumber' => $toNumber],
+                'playPrompt' => false,
+            ];
+
+            if ($callerId) {
+                $body['callerId'] = ['phoneNumber' => $fromNumber];
+            }
+
+            $response = $this->platform->post('/restapi/v1.0/account/~/extension/~/ring-out', $body);
+            $result = ['msg' => 'RingOut call status: ' . $response->json()->status->callStatus];
+            return json_encode($result);
+        } catch (Exception $e) {
+            $result = ['error' => 'RingOut Error: ' . $e->getMessage()];
+            return json_encode($result);
+        }
+    }
+
+    /**
+     * Get voicemail message content.
+     */
+    public function getVoicemailAttachment(string $messageId): string|false
+    {
+        $authErrorMsg = $this->authenticate();
+        if ($authErrorMsg !== 1) {
+            return text(js_escape($authErrorMsg));
+        }
+        if (empty($messageId)) {
+            return false; // Invalid message ID
+        }
+
+        try {
+            $endpoint = "/restapi/v1.0/account/~/extension/~/message-store/{$messageId}/content";
+            $response = $this->platform->get($endpoint);
+            return $response->raw(); // binary content for WAV/MP3
+        } catch (Exception $e) {
+            return false;
+        }
+    }
+
+    public function install()
+    {
+        try {
+            $token = $GLOBALS['ringcentral_voice_token'] ?? 'changeme';
+            if ($token == 'changeme') {
+                // Generate secure token
+                $token = bin2hex(random_bytes(16));
+                sqlStatement("REPLACE INTO `globals` (`gl_name`, `gl_index`, `gl_value`) VALUES ('ringcentral_voice_token', '', ?)", [$token]);
```

Comment:
Looks like storing a token in globals ringcentral_voice_token without a gui counterpart (ie. will not show up in Config). Let me know if this is wrong. If this is case, would store it somewhere else since this could mess up globals and would also be more secure if stored elsewhere (and would encrypt it).

Reply:
I could argue but I don't have the energy anymore. Anyway, you're our source code lifecycle guardian which I certainly respect so, will take care of.